### ISSUE TITLE
Lifetime of the clay object

### ISSUE THREAD

**Issue by emoon** [clay-ui-rs/clay#5]
So I want to be able to write code like this
```Rust
        clay.with([Layout::new()
                    .sizing_width(Sizing::Fixed(100.0))
                    .sizing_height(Sizing::Fixed(100.0))
                    .padding((10, 10))
                    .end(),
                Rectangle::new().color((255.0, 255.0, 255.0, 0.0)).end()], || {
                // some more code
                clay.with(...) {
                   ...
                }
                // do nothing
            },
        );
```
Currently `Clay` is single threaded, so it would be easy to solve with just having a single static function for `initialize()` that allocates the memory and such, but as I talked with Nick about it would be great if `Clay` wouldn't be single threaded and instead had it's own state so it doesn't modify global state.

Now with that said I think it makes sense to prepare for this to come so have a proper object for `Clay` is the way to go. One way would be to have something like a

```Rust
struct Clay {
  data: Rc<Vec<u8>>,
}
```
And then bump `Rc` when it goes down it goes down the levels. I'm not a big fan of this, but maybe that is the best way. Another way would be to have that we pass in a ref to clay instead in the closure.

so

```Rust
clay.with(..., |clay| {
   clay.with(...) // now uses ref passed in
}
```
Maybe this is better? I haven't tried it out yet if it will work correct.

**Comment by emoon** [comment]
Replying to myself. If I change to code to this
```Rust
    pub fn with<F: FnOnce(&Clay), const N: usize>(&self, configs: [TypedConfig; N], f: F) {
        unsafe { Clay__OpenElement() };

        for config in configs {
            if config.config_type == ElementConfigType::Id as _ {
                unsafe { Clay__AttachId(config.id) };
            } else if config.config_type == ElementConfigType::Layout as _ {
                unsafe { Clay__AttachLayoutConfig(config.config_memory as _) };
            } else {
                unsafe {
                    Clay__AttachElementConfig(
                        // This isn't strictcly correct, but as this is a union of pointers
                        // we can cast to any of them.
                        Clay_ElementConfigUnion {
                            rectangleElementConfig: config.config_memory as _,
                        },
                        config.config_type as _,
                    )
                };
            }
        }

        unsafe { Clay__ElementPostConfiguration() };

        f(self);

        unsafe { Clay__CloseElement(); }
    }
```
it seems to work fine, so one can write something like this
```Rust
        clay.with([Layout::new()
                    .sizing_width(Sizing::Fixed(100.0))
                    .sizing_height(Sizing::Fixed(100.0))
                    .padding((10, 10))
                    .end(),
                Rectangle::new().color((255.0, 255.0, 255.0, 0.0)).end()], |clay| 
            {
                clay.with([Layout::new()
                            .sizing_width(Sizing::Fixed(100.0))
                            .sizing_height(Sizing::Fixed(100.0))
                            .padding((10, 10))
                            .end(),
                Rectangle::new().color((255.0, 255.0, 255.0, 0.0)).end()], |_clay| 
                {

                }); 
            },
        );
```

**Comment by emoon** [comment]
My suggestion then is that I change the code to use a `Vec<u8>` internally, and we pass a ref down into the `with` closure.
